sqlmesh/core/engine_adapter/snowflake.py

```diff
@@ -204,9 +204,11 @@ def _df_to_source_queries(
             target_table or "pandas", quoted=False
         )  # write_pandas() re-quotes everything without checking if its already quoted
 
+        is_snowpark_dataframe = snowpark and isinstance(df, snowpark.dataframe.DataFrame)
```

Comment:
does this force 3.11 or is the snowpark check enough to allow python 3.12 if they don't use snowpark

Reply:
The latter. `snowpark` is defined as `optional_import("snowflake.snowpark")` which will return `False` on Python 3.12 so `is_snowpark_dataframe` will be False